### PR TITLE
Add Docker

### PR DESCRIPTION
**Pull request by ZamanovArslan**
### Summary

Add docker and docker-compose for development and CI.

Docker compose using two images: ruby2.6.6-alpine and postgres11-alpine

### How it works

N/A

### Test plan

For setup run `docker-compose up` this will build images and run the server on 0.0.0.0:3000

### Review notes

While reviewing pull-request (especially when it's your pull-request),
please make sure that:

- you understand what problem is solved by PR and how is it solved
- new tests are in place, no redundant tests
- DB schema changes reflect new migrations
- newly introduced DB fields have indexes and constraints
- there are no missed files (migrations, view templates)
- required ENV variables added and described in `.env.example` and added to Heroku
- associated Heroku review app works correctly with introduced changes

### Deploy notes

Notes regarding deployment the contained body of work.
These should note any db migrations, ENV variables, services, scripts, etc.

### References
* [GitHub Issue #15 ](https://github.com/fs/rails-base-graphql-api/issues/15)

### CODE REVIEW

File: .semaphore/semaphore.yml [hunk 1]
```diff
@@ -3,61 +3,105 @@ name: Rails Base GraphQL API
 
 agent:
   machine:
-    type: e1-standard-2
+    type: e1-standard-4
```
Comment (ArthurZaharov) [inline .semaphore/semaphore.yml:6]: Is there a reason to use more performant machines?

File: .semaphore/semaphore.yml [hunk 2]
```diff
@@ -3,61 +3,105 @@ name: Rails Base GraphQL API
...
+      secrets:
+        - name: docker-secrets
+
       prologue:
         commands:
-          - bundle install ${BUNDLE_ARGS} --local
-          - sem-service start postgres
-          - bin/rails db:setup
+          # Setup dynamic environment variables b/c they do not support via env_vars yet
+          - export BUNDLE_WITHOUT="development staging production"
```
Comment (ArthurZaharov) [inline .semaphore/semaphore.yml:67]: Can we use global_config prologue commands to remove duplication?

File: bin/setup [hunk 3]
```diff
@@ -7,6 +7,21 @@ if [ "$CI" ]; then
...
+# Fix Semaphore bug
+if [ "$SEMAPHORE" ]; then
+  sudo chown 1000:1000 -R ./
+fi
+
+# Build Docker image
+docker-compose up --build --detach
+
+# Set up database and add any development seed data
+docker-compose exec app bin/rails db:create db:schema:load
```
Comment (ArthurZaharov) [inline bin/setup:19]:
```suggestion
docker-compose exec app bin/rails db:setup
```

Should we use the setup script?

File: .semaphore/semaphore.yml [hunk 4]
```diff
@@ -3,61 +3,105 @@ name: Rails Base GraphQL API
...
+
+          # Authenticate with DockerHub
+          - echo "${DOCKER_PASSWORD}" | docker login -u "${DOCKER_USERNAME}" --password-stdin
+
+          # Pull image build on previous task
+          - docker pull "${WORKFLOW_CACHE_IMAGE}"
+          - docker tag "${WORKFLOW_CACHE_IMAGE}" rails_base_graphql_api:latest
+
+          # Setup dotenv
+          - cp .env.example .env
```
Comment (ArthurZaharov) [inline .semaphore/semaphore.yml:85]: Look like it is the same as bin/setup. Can we use it here?

File: .semaphore/semaphore.yml [hunk 5]
```diff
@@ -3,61 +3,105 @@ name: Rails Base GraphQL API
...
-        - name: Test
-          commands:
-            - bin/tests
+      - name: Check Running Images
+        commands:
+          - docker ps
+
+      - name: Run RSpec
+        commands:
+          - docker-compose exec app bin/rspec spec
```
Comment (ArthurZaharov) [inline .semaphore/semaphore.yml:103]: Let's leave bin/tests script here.

File: .semaphore/semaphore.yml [hunk 6]
```diff
@@ -3,61 +3,105 @@ name: Rails Base GraphQL API
...
-        - name: Quality
-          commands:
-            - bundle install ${BUNDLE_ARGS} --local
-            - bin/quality
+      - name: Docker build
+        commands:
+          - checkout
+
+          # Try to pull image for this branch
+          - docker pull "${CACHE_IMAGE}" || true
```
Comment (ArthurZaharov) [inline .semaphore/semaphore.yml:46]: If I understood correctly the image will be pulled after the initial commit pushed to GitHub. But what if in the next commits there will be changes in Gemfile? It means that we should rebuild the image. Is it correct?

File: .semaphore/semaphore.yml (hunk 1 above)
Comment (ArslanZamanov) [reply .semaphore/semaphore.yml:6]: I just copied that line from rails-base

Comment (ArthurZaharov) [reply .semaphore/semaphore.yml:6]: Did you compare the time on e1-standard-2 vs e1-standard-4?

File: .semaphore/semaphore.yml (hunk 6 above)
Comment (timurvafin) [reply .semaphore/semaphore.yml:46]: Lets try to play with this later, not sure

File: bin/setup (hunk 3 above)
Comment (timurvafin) [reply bin/setup:19]: Don’t think we should use setup script in tests since it will load unnecessary db/seeds in to the db

Comment (ArthurZaharov) [reply bin/setup:19]: Do we use [seedbank](https://github.com/james2m/seedbank) to separate env specific seeds?

File: .semaphore/semaphore.yml [hunk 7]
```diff
@@ -11,53 +11,83 @@ execution_time_limit:
...
+          # Setup database
+          - docker-compose exec app bin/rails db:create db:schema:load
 
       jobs:
-        - name: Test
-          commands:
-            - bin/tests
+      - name: Check Running Images
+        commands:
+          - docker ps
```
Comment (ArthurZaharov) [inline .semaphore/semaphore.yml:85]: I think that we should  remove this job

File: .semaphore/semaphore.yml [hunk 8]
```diff
@@ -11,53 +11,83 @@ execution_time_limit:
...
 
-  - name: Quality
-    task:
-      jobs:
-        - name: Quality
-          commands:
-            - bundle install ${BUNDLE_ARGS} --local
-            - bin/quality
+          # Build image
+          - docker-compose build
```
Comment (ArthurZaharov) [inline .semaphore/semaphore.yml:49]: It rebuilds the container each time. We should fix this.

File: .semaphore/semaphore.yml [hunk 9]
```diff
@@ -11,53 +11,83 @@ execution_time_limit:
...
+          - cp .env.example .env
+
+          # Fix Semaphore permissions bug
+          - sudo chown 1000:1000 -R ./
+
+          # Run container
+          - docker-compose up --detach
+
+          # Setup database
+          - docker-compose exec app bin/rails db:create db:schema:load
```
Comment (ArthurZaharov) [inline .semaphore/semaphore.yml:80]: We should not run this in the quality job.

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+# Add the Rails app
+COPY . /app/
+
+# Remove folders not needed in resulting image
+ARG FOLDERS_TO_REMOVE
+RUN rm -rf $FOLDERS_TO_REMOVE
+
+###############################
+# Stage Final
+FROM ruby:2.6.5-alpine
```
Comment (ArthurZaharov) [inline Dockerfile:39]: Please ensure that rebuild on semaphore using cache

File: bin/setup [hunk 11]
```diff
@@ -7,6 +7,21 @@ if [ "$CI" ]; then
   BUNDLER_ARGS="--without development staging production"
```
Comment (ArthurZaharov) [inline bin/setup:7]: We should rewrite the bin/setup to work only with docker

**Comment by DmitryBarskov** [comment]
Didn't work for MacOS
[asciicast](https://asciinema.org/a/335025)

**Comment by ArthurZaharov** [comment]
@timurvafin Could you review this PR. It looks good for me.

**Pull request merged**